numpy/_core/src/multiarray/stringdtype/dtype.c

```diff
@@ -0,0 +1,910 @@
+ * (This should get a bit nicer eventually with Python >3.11.)
+ */
+PyArray_DTypeMeta PyArray_StringDType = {
+        {{
+                PyVarObject_HEAD_INIT(NULL, 0).tp_name =
+                        "numpy.dtypes.StringDType",
+                .tp_basicsize = sizeof(PyArray_StringDTypeObject),
+                .tp_new = stringdtype_new,
+                .tp_dealloc = (destructor)stringdtype_dealloc,
+                .tp_repr = (reprfunc)stringdtype_repr,
+                .tp_str = (reprfunc)stringdtype_repr,
+                .tp_methods = PyArray_StringDType_methods,
+                .tp_members = PyArray_StringDType_members,
+                .tp_richcompare = PyArray_StringDType_richcompare,
+                .tp_hash = PyArray_StringDType_hash,
+        }},
+        /* rest, filled in during DTypeMeta initialization */
+};
+
+NPY_NO_EXPORT int
+init_string_dtype(void)
+{
+    PyArrayMethod_Spec **PyArray_StringDType_casts = get_casts();
+
+    PyArrayDTypeMeta_Spec PyArray_StringDType_DTypeSpec = {
+            .flags = NPY_DT_PARAMETRIC,
+            .typeobj = &PyUnicode_Type,
+            .slots = PyArray_StringDType_Slots,
+            .casts = PyArray_StringDType_casts,
+    };
+
+    /* Loaded dynamically, so may need to be set here: */
+    ((PyObject *)&PyArray_StringDType)->ob_type = &PyArrayDTypeMeta_Type;
+    ((PyTypeObject *)&PyArray_StringDType)->tp_base = &PyArrayDescr_Type;
+    if (PyType_Ready((PyTypeObject *)&PyArray_StringDType) < 0) {
+        return -1;
+    }
+
+    if (dtypemeta_initialize_struct_from_spec(
+                &PyArray_StringDType, &PyArray_StringDType_DTypeSpec, 1) < 0) {
+        return -1;
+    }
+
+    PyArray_Descr *singleton =
+            NPY_DT_CALL_default_descr(&PyArray_StringDType);
+
+    if (singleton == NULL) {
+        return -1;
+    }
+
+    PyArray_StringDType.singleton = singleton;
+    PyArray_StringDType.type_num = NPY_VSTRING;
+
+    for (int i = 0; PyArray_StringDType_casts[i] != NULL; i++) {
+        PyMem_Free(PyArray_StringDType_casts[i]->dtypes);
+        PyMem_Free(PyArray_StringDType_casts[i]);
+    }
+
+    PyMem_Free(PyArray_StringDType_casts);
+
+    return 0;
+}
+
+int
+free_and_copy(npy_string_allocator *in_allocator,
+              npy_string_allocator *out_allocator,
+              const npy_packed_static_string *in,
+              npy_packed_static_string *out, const char *location)
+{
+    if (NpyString_free(out, out_allocator) < 0) {
+        npy_gil_error(PyExc_MemoryError, "Failed to deallocate string in %s", location);
+        return -1;
+    }
+    if (NpyString_dup(in, out, in_allocator, out_allocator) < 0) {
+        npy_gil_error(PyExc_MemoryError, "Failed to allocate string in %s", location);
+        return -1;
+    }
+    return 0;
+}
+
+NPY_NO_EXPORT int
+load_new_string(npy_packed_static_string *out, npy_static_string *out_ss,
+                size_t num_bytes, npy_string_allocator *allocator,
+                const char *err_context)
+{
```

Comment:
This one and `NpyString_load` purposely have this API because I want to encourage the pattern of stack-allocating `out_ss` in the caller. Copied the comment from `NpyString_load` over here to make that clearer. My original intent was to make `NpyString_load` public but eventually decided to only make the more limited C API public, that's why it's a lot more polished than `load_new_string` which was originally intended only as an internal helper.

Reply:
Yes, that makes sense, and indeed it became clear once I got further in the review. But good to have a comment for people who see the function in isolation, as I did.